Include/pyhash.h

```diff
@@ -12,8 +12,27 @@ PyAPI_FUNC(Py_hash_t) _Py_HashPointer(void*);
 PyAPI_FUNC(Py_hash_t) _Py_HashBytes(const void*, Py_ssize_t);
 #endif
 
-/* Prime multiplier used in string and various other hashes. */
-#define _PyHASH_MULTIPLIER 1000003UL  /* 0xf4243 */
+/* Multiplier used for various hashes */
+#define _PyHASH_MULTIPLIER 1000003UL
+
+/* Official constants from the xxHash specification. Optimizing compilers
+ * should emit a single "rotate" instruction for the _PyHASH_XXROTATE()
+ * expansion. If that doesn't happen for some important platform, the macro
+ * could be changed to expand to a platform-specific rotate spelling instead.
+ */
+#if SIZEOF_PY_UHASH_T > 4
+#define _PyHASH_XXPRIME_1 ((Py_uhash_t)11400714785074694791ULL)
+#define _PyHASH_XXPRIME_2 ((Py_uhash_t)14029467366897019727ULL)
+#define _PyHASH_XXPRIME_3 ((Py_uhash_t)1609587929392839161ULL)
+#define _PyHASH_XXPRIME_5 ((Py_uhash_t)2870177450012600261ULL)
+#define _PyHASH_XXROTATE(x) ((x << 31) | (x >> 33))  /* Rotate left 31 bits */
+#else
+#define _PyHASH_XXPRIME_1 ((Py_uhash_t)2654435761UL)
+#define _PyHASH_XXPRIME_2 ((Py_uhash_t)2246822519UL)
+#define _PyHASH_XXPRIME_3 ((Py_uhash_t)3266489917UL)
+#define _PyHASH_XXPRIME_5 ((Py_uhash_t)374761393UL)
+#define _PyHASH_XXROTATE(x) ((x << 13) | (x >> 19))  /* Rotate left 13 bits */
+#endif
```

Comment:
Please move this into tupleobject.c to keep it all self-contained and no leak outside of the world of tuple objects.

Reply:
> Please move this into tupleobject.c to keep it all self-contained and no leak outside of the world of tuple objects.

Why do you consider this a "leak"? I think it's good to define that in the header file: it's not a priori meant to be specific to tuples. I can easily imagine other use cases for this new hash.

Similarly, this is also in the header file because it's used by more than one hash implementation:
```
#define _PyHASH_MULTIPLIER 1000003UL  /* 0xf4243 */
```

Reply:
I agree with Raymond here.  The difference is that `_PyHASH_MULTIPLIER` _is_ used in other places now.  The `XX` defines are not.  If other modules want to move toward an xxHash approach, that's the time to refactor - and would likely include moving the implementation code as well as the defines.  It doesn't really help to have the defines in a globally visible place but the implementation code visible only in `tupleobject.c`.  The "implementation code" for ` _PyHASH_MULTIPLIER` consists of merely sticking `*` in front of it.

Note that the xxHash implementation I posted to BPO _did_ keep it all inside `tupleobject.c`.  For the same reasons.  Pretending to generality that isn't actually there is an anti-pattern.